### DISCUSSION TITLE
Computation of functionals over arrays?

### DISCUSSION THREAD

**Discussion by ThoChat** [javicarron/pynkowski#6] [Q&A]
Hi there!
I'm interested in computing Minkowski functionals on 2D arrays. Since I'm only interested in binary arrays I'm currently using the QuantImPy library however, since this is not maintained anymore I was wondering if you were considering to implement computation from arrays ? 
From your doc it seems that this should arrive soon. Is it possible to get any updates about this? 
Thanks for the time you take to reply and for sharing your amazing work!

Cheers

**Comment by javicarron** [comment]
Hello and thanks for the interest!

It should be possible now, for numpy arrays of arbitrary dimension. Keep in mind that it has not been fully tested, and some functionality regarding the theoretical predictions is not fully implemented. If you have any problems at all, don't hesitate to open an issue.

I have just tested this code:
```import numpy as np
import matplotlib.pyplot as plt
from scipy.ndimage import gaussian_filter
import pynkowski as mf


# Create a random 2D field
square_map = np.random.randn(128, 128)
smoothed = gaussian_filter(square_map, 5)
smoothed -= smoothed.mean()                            # To compare with the theoretical prediction

# Thresholds for the excursion set, in units of the standard deviation
us = np.linspace(-3, 3, 100)                           

image = mf.data.DataArray(smoothed, normalise=True)    # Normalise=True to have unit variance (but this does not change the mean)

# MFs on the image
v0d = mf.V0(image, us)
v1d = mf.V1(image, us)
v2d = mf.V2(image, us)

# Theoretical predictions
mu = (image.first_der[0].var() + image.first_der[1].var()) / 2
theo = mf.Gaussian(image.field.ndim, 1, mu)            # The 1 is the standard deviation of the field

v0t = mf.V0(theo, us)
v1t = mf.V1(theo, us)
v2t = mf.V2(theo, us)


# Plotting
plt.figure(figsize=(25, 5))
plt.subplot(131)
plt.plot(us, v0d, label='Data')
plt.plot(us, v0t, label='Theory')
plt.xlabel('Threshold')
plt.ylabel('V0')
plt.legend()
plt.subplot(132)
plt.plot(us, v1d, label='Data')
plt.plot(us, v1t, label='Theory')
plt.xlabel('Threshold')
plt.ylabel('V1')
plt.legend()
plt.subplot(133)
plt.plot(us, v2d, label='Data')
plt.plot(us, v2t, label='Theory')
plt.xlabel('Threshold')
plt.ylabel('V2')
plt.legend()

```

which produced this plot:
![imagen](https://github.com/javicarron/pynkowski/assets/25009611/37447e60-f76b-4bbe-94f2-c1e84016656a)


You can find relevant documentation at https://javicarron.github.io/pynkowski/pynkowski/data/array.html and https://javicarron.github.io/pynkowski/pynkowski/stats/minkowski.html

Let us know if it works :smile:

**Comment by javicarron** [comment]
I forgot to mention, for binary fields your Minkowski Functionals should look like a double step function (you have the excursion set is the whole space for $u<0$, equal to the regions where the field is 1 for $0 < u < 1$, and empty when $u>1$). For these fields you may find some numerical instabilities. Slightly smoothing the field may help with that.
